README

		Tcl/Tk Image formats (TkImg)
		(Version 1.4 / September 2010)

		Jan Nijtmans, Andreas Kupries, Paul Obermeier

This file contains a collection of format handlers for the Tk photo
image type, and a new image type, pixmaps.  It can be used in
combination with Tcl/Tk 8.3 or later but 8.5.9 (or 8.6, when
available) are highly recommended.

Included in this distribution are the most recent versions (as of
sept 2010) of the libz, libpng, libjpeg, and libtiff libraries.
These are not required, unless you need support for the PNG, JPEG,
or TIFF format. Note that you have to build these libraries to
support the named formats, even if your system already has shared
libraries for these formats. This is because the libraries here are
built such that they can be loaded as packages by the Tcl/Tk core,
making the handling of the various dependencies much easier. An
earlier version, 1.2.4, used a modified copy of Tcl's functions for
loading of shared libraries to load the support libraries at runtime.
These have been abandoned in favor of the new approach.

BACKGROUND TO BUILDING

The TkImg sources have been heavily refactored. Instead of building
one big package there are several small packages most of which are
independent of each other.

All packages employ a TEA 2 based build system (configure/Make) for
easy configuration and building. This means that 99% of the options
to configure are the same for all the packages.

All packages save for the supporting ones (see below) have names
starting with 'img'. For compatibility with older scripts the
package index file of the utility package 'img::base' declares
pseudo-packages 'img' and 'Img' which load all packages provided
by this extension.

BUILDING - The convenient way

The toplevel directory of tkImg contains a configure and Makefile for
convenient configuration and building of all packages in the
extension.

Run this configure script in a build directory of your choice to
configure all packages, then run

	make all
	make install

in the build directory. The most important options for configure are
--prefix and --exec-prefix which tell the build system where to find
Tcl and Tk, and also where to install the results of the builds
(header, shared libraries, ...).

This convenient build installs all shared libraries in a single
directory 'Img<VERSION>, together with a union package index for all
packages.

If the builder requires only a subset of the packages he can use

	make all

to build all packages, but must not use 'make install' in the top
build directory. Instead run

	make install

in the sub directories of the build directory which contain the
requested packages, and their dependencies !

Note, that target install also builds the documentation based on 
Tcl script dtplite. To build and install without documentation, use

	make install-libraries

PACKAGES AND THEIR DEPENDENCIES

This section is for people who want to bypass the master
configure/Makefile.

At the bottom of the hierarchy of dependencies are the 'support
packages'. They encapsulate the support libraries for ZIP, TIFF, PNG
and JPEG in packages directly loadable by Tcl. They provide the
functionality of their underlying libraries via stub-tables. The names
of these packages are:

	Name		Directory		    Dependency
	-------------------------------------------------
	zlibtcl		tkimg/compat/libz
	pngtcl		tkimg/compat/libpng	  zlibtcl
	jpegtcl		tkimg/compat/libjpeg
	tifftcl		tkimg/compat/libtiff
	-------------------------------------------------

The package 'pngtcl' depends on 'zlibtcl' and has to be built after
it. Beyond that the order of building is free. Their TEA build systems
can be found in the directories listed above.

The configure/Makefile's in the directories

	tkimg/compat/libz, tkimg/compat/libpng, tkimg/compat/libjpeg, tkimg/compat/libtiff

must _not_ be used directly under any circumstances.


At the next level is the package 'tkimg'. It provides utility
functions used by all format handlers, and therefore has to be built
before any of them. Its TEA build system can be found in the directory

	Name		Directory		Dependency
	-------------------------------------------------
	img::base	tkimg/base
	-------------------------------------------------

Beyond that are the packages providing the various format
handlers. They can be built in any order. Some of them depend on
supporting packages, as noted below. Their names and directories are

	Name		Directory	Dependency
	-------------------------------------------------
	img::bmp	tkimg/bmp
	img::dted	tkimg/dted
	img::gif	tkimg/gif
	img::ico	tkimg/ico
	img::jpeg	tkimg/jpeg	jpegtcl
	img::pcx	tkimg/pcx
	img::pixmap	tkimg/pixmap
	img::png	tkimg/png	pngtcl, zlibtcl
	img::ppm	tkimg/ppm
	img::ps		tkimg/ps
	img::raw	tkimg/raw
	img::sgi	tkimg/sgi
	img::sun	tkimg/sun
	img::tga	tkimg/tga
	img::tiff	tkimg/tiff	tifftcl, zlibtcl, jpegtcl
	img::window	tkimg/window
	img::xbm	tkimg/xbm
	img::xpm	tkimg/xpm
	-------------------------------------------------

CONFIGURE

The source distributions have "configure", which  has the following
possibilities:
	configure ?options?

Some possible options:
	--prefix=<dir>		Install headers in <dir>/include.
				Default: "/usr/local"
	--exec-prefix=<dir>	Install libraries in <dir>/lib.
				Default: "${prefix}"
	--with-tcl=<dir>	Find Tcl libraries in <dir> and include files in
				<dir>/../include or <dir>/../generic.
				Default: "${exec-prefix}/lib" or a sibling directory
				of the Img build directory.
	--with-tk=<dir>		Find Tk libraries in <dir> and include files in
				<dir>/../include or <dir>/../generic.
				Default: the same values as given by --with-tcl
				or a sibling directory of the Img build directory.

	Packages dependent on one or more of the support packages have
	appropriate options to locate their libraries and headers, if
	they cannot be found via --prefix and --exec-prefix

	Use configure --help to query each configure for the full set
	of options available to it.


MAKE & INSTALL

	'make', or 'make all' will build any package.

	'make install' will install its headers and libraries in the
	directories specified via --prefix/include and --exec-prefix/lib.

USE

If you want to use one or more of new image formats to be available in
your own scripts, add a line like the following somewhere in your
script:

	package require img::FORMAT

where FORMAT stands for the name of the format requested. See the list
in section 'PACKAGES AND THEIR DEPENDENCIES' for the available
formats. Each format will automatically load all of the packages it is
dependent on.

The line
	package require Img

is still legal and will load all format handlers provided by this
extension except dted and raw.

As an example, have a look at the file "demo.tcl"

For further options see the manpages.


DOCUMENTATION

The documentation coming with the distribution is written in the
doctools format (see tcllib).

	NOTE:
		The Img documentation makes use of features (include
		files and variables) which are not part of an official
		tcllib release yet. A cvs snapshot of tcllib is
		required.



PLATFORMS
	tested:
		Linux 2.0.30	(gcc, egcs)
		IRIX 6.2	(cc, 32 bits only)
		ULTRIX 4.1	(cc, gcc, egcs)
		ULTRIX 4.4	(cc 3.0)
		HP-UX-9		(cc, gcc, egcs)
		HP-UX-10.2	(gcc, egcs)
		SunOS-4.1.2     (cc)
		Solaris 2.5	(gcc, egcs)
		FreeBSD	2.2	(gcc, egcs)
		Windows	95/98/NT(gcc-2.95/MS Visual C++ 4.2/5.0/6.0)

	Not done yet:
		Mac		) someone interested
		VMS		) to try??????

	Other machines and OS's should work too. 

Feedback about other OS's and compilers is appreciated. Please
help me in making this package better.

--
		Jan Nijtmans
		email: nijtmans@users.sourceforge.net
		url:   http://purl.oclc.org/net/nijtmans/

		Andreas Kupries, ActiveState Corp
		email:	andreas_kupries@users.sourceforge.net	(private)
			akupries@shaw.ca			(private)
			andreask@activestate.com		(work)
		url:	http://www.purl.org/NET/akupries/


ChangeLog
2010-09-17  Jan Nijtmans <nijtmans@users.sourceforge.net>

	*
	* Released and tagged Tkimg 1.4 ========================
	*

2010-08-30  Jan Nijtmans <nijtmans@users.sourceforge.net>

	* license.terms: Fix DFARs note for number-adjusted rights clause

2010-07-06  Jan Nijtmans <nijtmans@users.sourceforge.net>

	* genStubs.tcl: "generic" argument is no longer necessary when
	there are no suppressors.
	* [tkimg-Bugs-3011001] Only support accessing tkImg
	procs through stub table implemented for libpng

2010-06-30  Jan Nijtmans <nijtmans@users.sourceforge.net>

	* upgrade to libtiff 3.9.4

2010-06-28  Jan Nijtmans <nijtmans@users.sourceforge.net>

	* upgrade to libpng 1.4.3
	* backport various ttkGenStubs.tcl improvements
	  to genStubs.tcl

2010-06-17  Jan Nijtmans <nijtmans@users.sourceforge.net>

	* upgrade to libjpeg version 8b

2010-06-04  Jan Nijtmans <nijtmans@users.sourceforge.net>

	* [tkimg-Bugs-3011001] Only support accessing tkImg
	procs through stub table implemented for base and zlib

2010-05-11  Jan Nijtmans <nijtmans@users.sourceforge.net>

	* upgrade to libpng 1.4.2

2010-05-04  Jan Nijtmans <nijtmans@users.sourceforge.net>

	* Make type "boolean" available everywhere, so individual
	  handlers don't need to define Boln any more.
	* Formatting.
	* Make --enable-symbols build on MSVC a little bit
	  closer to reality and explain disabled warnings.

2010-05-03  Jan Nijtmans <nijtmans@users.sourceforge.net>

	* Use PNGTCLAPI in stead of EXTERN in tkImg png package
	* xbm fix unused variable

2010-04-30  Andreas Kupries <andreask@activestate.com>

	* Fix building tkimg --disable-shared

2010-04-28  Jan Nijtmans <nijtmans@users.sourceforge.net>

	* More MODULE_SCOPE functions in base
	* Use TKIMGAPI in stead of EXTERN in tkImg base package
	* Fix various gcc and MSVC warnings
	* upgrade to libpng 1.4.1

2010-04-23  Jan Nijtmans <nijtmans@users.sourceforge.net>

	* upgrade to zlib 1.2.5

2010-04-14  Jan Nijtmans <nijtmans@users.sourceforge.net>

	* Regenerate all configure scripts with latest TEA tcl.m4

2010-04-09  Jan Nijtmans <nijtmans@users.sourceforge.net>

	* Only build pkgIndex.tcl when something changed.

2010-04-07  Jan Nijtmans <nijtmans@users.sourceforge.net>

	* TkimgInitUtilities() always succeeds, and does not
	  return TCL_ERROR/TCL_OK.
	* Eliminate forced dependancy for aqua on
	  ::tk::unsupported::MacWindowStyle
	* Eliminate msgcat compatibility layer

2010-04-01  Jan Nijtmans <nijtmans@users.sourceforge.net>

	* [Bug 2975449] Kill the Tk 8.2/8.3 ifdeffery - at
	  least the runtime swizzle (I don't consider
	  this a bug at all, because it worked fine
	  as long as it was needed!)
	* [Bug 2907616] configure ignores --mandir switch

2010-03-31  Jan Nijtmans <nijtmans@users.sourceforge.net>

	* Sync with latest TEA (tcl.m4), regenerated all
	  configure scripts with autoconf-2.65.
	* Sync genStubs.tcl with latest one from Tcl 8.6,
	  regenerated all *Decls.h files. All stubs
	  pointers are const now.
    
2009-07-26  Jan Nijtmans <nijtmans@users.sourceforge.net>

	* Sync with latest TEA (tcl.m4), regenerated all
	  configure scripts.
    
2009-04-28  Jan Nijtmans <nijtmans@users.sourceforge.net>

	* Fix [Bug #82910] teacup requires libX11 to run
    
2009-03-12  Stuart Cassoff <stwo@users.sourceforge.net>

	* various fixes to tcl.m4 [Bug #2669308] (Stuart Cassoff)
	  (To be checked in in sampleextension as well.)
	* fix [Bug #2669482] Can't 'make test'

2009-03-12  Jan Nijtmans <nijtmans@users.sourceforge.net>

	* fix "make checkstubs"
	* a few CONST84 eliminations

2009-03-09  Jan Nijtmans <nijtmans@users.sourceforge.net>

	* Revert to using "ld -b" in stead of "${CC} -b" for
	  HP-UX-11; it doesn't work in 64-bit mode.

2009-03-03  Jan Nijtmans <nijtmans@users.sourceforge.net>

	* fix "make genstubs" for libpng, libtiff, libz, libjpeg
	* PNG build fix

2009-02-24  Jan Nijtmans <nijtmans@users.sourceforge.net>

	* gif handler cleanup
	* Allow -index N option for TIFF images too [Feature Request #706277]

2009-02-23  Jan Nijtmans <nijtmans@users.sourceforge.net>

	* Attempt to fix HP-UX Itanium 64-bit build
	* revert r177 change to pkgIndex.tcl: it broke TEApot package
	  handling.

2009-02-17  Jan Nijtmans <nijtmans@users.sourceforge.net>

	* Move ./libz to ./compat/zlib and ./libz/tcl to ./zlib
	* Move ./libjpeg to ./compat/libjpeg and ./libjpeg/tcl to ./libjpeg
	* Move ./libpng to ./compat/libpng and ./libpng/tcl to ./libpng
	* Move ./libtiff to ./compat/libtiff and ./libtiff/tcl to ./libtiff
	* Require Tcl/Tk 8.3 for libz, libpng, libtiff and libjpeg

2009-01-12  Jan Nijtmans <nijtmans@users.sourceforge.net>

	* upgrade to libpng 1.2.34

2009-01-12	Jan Nijtmans <nijtmans@users.sourceforge.net>

	* No longer Panic when an out-of-memory error occurs,
	  when compiled with Tcl/Tk 8.5 or later and running with
	  Tcl/Tk 8.5 and later (Tk bug #2477446)

2009-01-05  Jan Nijtmans <nijtmans@users.sourceforge.net>

	* Two more stub functions for improved Tcl 8.5 support.
	* Make sure that in C preprocessor directives the "#"
	  character is the first character on the line.

2008-11-13  Jan Nijtmans <nijtmans@users.sourceforge.net>

	* For gcc, from now on use as standard CFLAGS:
	    CFLAGS_OPTIMIZE="-O2 -fomit-frame-pointer"
	    CFLAGS_WARNING="-Wall -Wwrite-strings"
	* Upgrade to GNU Autoconf 2.63

2008-11-13  Jan Nijtmans <nijtmans@users.sourceforge.net>

	* Create 1.3.x branch. Now trunk is available for
	  tkImg 1.4 development.

2008-11-13  Andreas Kupries <andreask@activestate.com>

	* Fix win32 compile issue with zlib, mask gzprintf
    
2008-11-04	Paul Obermeier <obermeier@users.sourceforge.net>

	* Added new (visual) test suite for checking the different
	  ways of reading and writing images in the various formats.

2008-11-04	Paul Obermeier <obermeier@users.sourceforge.net>

	* Updated demo program to include all available image
	  formats. Also added additional tests for all image formats.

2008-11-04	Paul Obermeier <obermeier@users.sourceforge.net>


ps/ps.c
/*
 * ps.c --
 *
 *  PS + PDF photo image type, Tcl/Tk package
 *
 * Author : Jan Nijtmans *
 * Date   : 7/24/97        *
 *
 * Copyright (c) 2002 Andreas Kupries <andreas_kupries@users.sourceforge.net>
 *
 * $Id: ps.c,v 1.2 2004/06/03 21:54:43 andreas_kupries Exp $
 *
 */

/*
 * Generic initialization code, parameterized via CPACKAGE and PACKAGE.
 */

/*
 * Activate second 'format' definition in init.c
 * share all functions except for the matching
 */
#define SECOND_FORMAT
#define SECOND_CHNREAD	ChnRead
#define SECOND_OBJREAD	ObjRead
#define SECOND_CHNWRITE	ChnWrite
#define SECOND_STRWRITE	StringWrite

#include "init.c"


#include <string.h>
#include <stdlib.h>

/*
 * Prototypes for local procedures defined in this file:
 */

static int CommonMatchPS _ANSI_ARGS_((tkimg_MFile *handle, Tcl_Obj *format,
	int *widthPtr, int *heightPtr));

static int CommonMatchPDF _ANSI_ARGS_((tkimg_MFile *handle, Tcl_Obj *format,
	int *widthPtr, int *heightPtr));

static int parseFormat _ANSI_ARGS_((Tcl_Obj *format, int *zoomx,
	int *zoomy));

static int CommonRead _ANSI_ARGS_((Tcl_Interp *interp, tkimg_MFile *handle,
	Tcl_Obj *format, Tk_PhotoHandle imageHandle, int destX, int destY,
	int width, int height, int srcX, int srcY));

static int CommonWrite _ANSI_ARGS_((Tcl_Interp *interp, tkimg_MFile *handle,
	Tcl_Obj *format, Tk_PhotoImageBlock *blockPtr));



static int
parseFormat(format, zoomx, zoomy)
     Tcl_Obj *format;
     int *zoomx;
     int *zoomy;
{
    int objc, i, length, index = 0;
    Tcl_Obj **objv = NULL;
    char *p;
    double zx = 1.0, zy = 1.0;

    if (!format) {
	*zoomx = (int) (72 * zx + 0.5);
	*zoomy = (int) (72 * zy + 0.5);
    }

    if (tkimg_ListObjGetElements((Tcl_Interp*) NULL, format, &objc, &objv) != TCL_OK) {
	return -1;
    }
    for (i=1; i<objc; i++) {
	p = Tcl_GetStringFromObj(objv[i], &length);
	if ((p[0] == '-') && ((i+1)<objc)) {
	    if (length < 2) {
		index = -1; break;
	    }
	    if (!strncmp(p,"-index", length)) {
		if (Tcl_GetIntFromObj((Tcl_Interp *) NULL, objv[++i], &index) != TCL_OK) {
		    index = -1; break;
		}
	    } else if (!strncmp(p, "-zoom", length)) {
		if (Tcl_GetDoubleFromObj((Tcl_Interp *) NULL, objv[++i], &zx) != TCL_OK) {
		    index = -1; break;
		}
		if (i > objc) {
		    zy = zx;
		} else {
		    p = Tcl_GetStringFromObj(objv[i+1], &length);
		    if (p[0] != '-') {
			if (Tcl_GetDoubleFromObj((Tcl_Interp *) NULL, objv[++i], &zy) != TCL_OK) {
			    index = -1; break;
			}
		    } else {
			zy = zx;
		    }
		}
	    } else {
		index = -1; break;
	    }
	} else {
	    if (Tcl_GetIntFromObj((Tcl_Interp *) NULL, objv[++i], &index) != TCL_OK) {
		index = -1; break;
	    }
	}
    }
    if (!index) {
	*zoomx = (int) (72 * zx + 0.5);
	*zoomy = (int) (72 * zy + 0.5);
    }
    return index;
}

static int
ChnMatch(interp, chan, fileName, format, widthPtr, heightPtr)
    Tcl_Interp *interp;
    Tcl_Channel chan;
    CONST char *fileName;
    Tcl_Obj *format;
    int *widthPtr, *heightPtr;
{
    tkimg_MFile handle;

    tkimg_FixChanMatchProc(&interp, &chan, &fileName, &format, &widthPtr, &heightPtr);

    handle.data = (char *) chan;
    handle.state = IMG_CHAN;

    return CommonMatchPS(&handle, format, widthPtr, heightPtr);
}

static int
ObjMatch(interp, data, format, widthPtr, heightPtr)
    Tcl_Interp *interp;
    Tcl_Obj *data;
    Tcl_Obj *format;
    int *widthPtr, *heightPtr;
{
    tkimg_MFile handle;

    tkimg_FixObjMatchProc(&interp, &data, &format, &widthPtr, &heightPtr);

    handle.data = tkimg_GetStringFromObj(data, &handle.length);
    handle.state = IMG_STRING;

    return CommonMatchPS(&handle, format, widthPtr, heightPtr);
}

static int
CommonMatchPS(handle, format, widthPtr, heightPtr)
    tkimg_MFile *handle;
    Tcl_Obj *format;
    int *widthPtr, *heightPtr;
{
    unsigned char buf[41];

    if ((tkimg_Read(handle, (char *) buf, 11) != 11)
	    || (strncmp("%!PS-Adobe-", (char *) buf, 11) != 0)) {
	return 0;
    }
    while (tkimg_Read(handle,(char *) buf, 1) == 1) {
	if (buf[0] == '%' &&
		(tkimg_Read(handle, (char *) buf, 2) == 2) &&
		(!memcmp(buf, "%B", 2) &&
		(tkimg_Read(handle, (char *) buf, 11) == 11) &&
		(!memcmp(buf, "oundingBox:", 11)) &&
		(tkimg_Read(handle, (char *) buf, 40) == 40))) {
	    int w, h, zoomx, zoomy;
	    char *p = buf;
	    buf[41] = 0;
	    w = - (int) strtoul(p, &p, 0);
	    h = - (int) strtoul(p, &p, 0);
	    w += strtoul(p, &p, 0);
	    h += strtoul(p, &p, 0);
	    if (parseFormat(format, &zoomx, &zoomy) >= 0) {
		w = (w * zoomx + 36) / 72;
		h = (h * zoomy + 36) / 72;
	    }
	    if ((w <= 0) || (h <= 0)) return 0;
	    *widthPtr = w;
	    *heightPtr = h;
	    return 1;
	}
    }
    return 0;
}

static int
ChnRead(interp, chan, fileName, format, imageHandle,
	destX, destY, width, height, srcX, srcY)
    Tcl_Interp *interp;
    Tcl_Channel chan;
    CONST char *fileName;
    Tcl_Obj *format;
    Tk_PhotoHandle imageHandle;
    int destX, destY;
    int width, height;
    int srcX, srcY;
{
    tkimg_MFile handle;

    handle.data = (char *) chan;
    handle.state = IMG_CHAN;

    return CommonRead(interp, &handle, format, imageHandle, destX, destY,
	    width, height, srcX, srcY);
}

static int
ObjRead(interp, data, format, imageHandle,
	destX, destY, width, height, srcX, srcY)
    Tcl_Interp *interp;
    Tcl_Obj *data;
    Tcl_Obj *format;
    Tk_PhotoHandle imageHandle;
    int destX, destY;
    int width, height;
    int srcX, srcY;
{
    tkimg_MFile handle;

    tkimg_ReadInit(data,'%',&handle);

    return CommonRead(interp, &handle, format, imageHandle, 
	    destX, destY, width, height, srcX, srcY);
}

typedef struct myblock {
    Tk_PhotoImageBlock ck;
    int dummy; /* extra space for offset[3], in case it is not
		  included already in Tk_PhotoImageBlock */
} myblock;

#define block bl.ck

static int
CommonRead(interp, handle, format, imageHandle,
	destX, destY, width, height, srcX, srcY)
    Tcl_Interp *interp;
    tkimg_MFile *handle;
    Tcl_Obj *format;
    Tk_PhotoHandle imageHandle;
    int destX, destY;
    int width, height;
    int srcX, srcY;
{
#ifndef MAC_TCL
    char *argv[10];
    int len, i, j, fileWidth, fileHeight, maxintensity, index;
    char *p, type;
    unsigned char buffer[1025], *line = NULL, *line3 = NULL;
	char zoom[64], papersize[64];
    Tcl_Channel chan;
    Tcl_DString dstring;
    myblock bl;
    int zoomx, zoomy;

    index = parseFormat(format, &zoomx, &zoomy);
    if (index < 0) {
	Tcl_AppendResult(interp, "invalid format: \"",
		tkimg_GetStringFromObj(format, NULL), "\"", (char *) NULL);
	return TCL_ERROR;
    }
    sprintf(zoom, "-r%dx%d", zoomx, zoomy);

    len = tkimg_Read(handle, buffer, 1024);
    buffer[1024] = 0;
    p = strstr(buffer,"%%BoundingBox:");
    fileHeight = height + srcY;
    if (p) {
	/* postscript */
	p += 14;
	srcX += (strtoul(p, &p, 0) * zoomx + 36) / 72;
	fileHeight += (strtoul(p, &p, 0) * zoomy + 36) / 72;
	strtoul(p, &p, 0);
	srcY -= (strtoul(p, &p, 0) * zoomy + 36) / 72;
    } else {
	/* pdf */

	/*
	 * Extract the pixel position of the upper left corner
	 * of the image from the file. How to do that????
	 * For now I just assume A4-size with 72 pixels/inch.
	 */
	srcX += (0 * zoomx + 36) / 72;
	srcY -= (792 * zoomy + 36) /72;
    }

    sprintf(papersize, "-g%dx%d", srcX+width, fileHeight);

    argv[0] = "gs";
    argv[1] = "-sDEVICE=ppmraw";
    argv[2] = zoom;
    argv[3] = papersize;
    argv[4] = "-q";
    argv[5] = "-dNOPAUSE";
    argv[6] = "-sOutputFile=-";
    argv[7] = "-";

    chan = Tcl_OpenCommandChannel(interp, 8, argv,
	    TCL_STDIN|TCL_STDOUT|TCL_STDERR|TCL_ENFORCE_MODE);
    if (!chan) {
	return TCL_ERROR;
    }
    if (Tcl_SetChannelOption(interp, chan, "-translation", "binary") != TCL_OK) {
	return TCL_ERROR;
    }

    while (len > 0) {
	Tcl_Write(chan, (char *) buffer, 1024);
	len = tkimg_Read(handle, buffer, 1024);
    }
    Tcl_Write(chan,"\nquit\n", 6);
    Tcl_Flush(chan);

    Tcl_DStringInit(&dstring);
    len = Tcl_Gets(chan, &dstring);
    p = Tcl_DStringValue(&dstring);
    type = p[1];
    if ((p[0] != 'P') || (type < '4') || (type > '6')) {
	Tcl_AppendResult(interp, "gs error: \"",
		p, "\"",(char *) NULL);
	return TCL_ERROR;
    }
    do {
	Tcl_DStringSetLength(&dstring, 0);
	Tcl_Gets(chan, &dstring);
	p = Tcl_DStringValue(&dstring);
    } while (p[0] == '#');
    fileWidth = strtoul(p, &p, 0);
    srcY += (fileHeight = strtoul(p, &p, 0));

    if ((srcX + width) > fileWidth) {
	width = fileWidth - srcX;
    }
    if ((srcY + height) > fileHeight) {
	height = fileHeight - srcY;
    }
    if ((width <= 0) || (height <= 0)) {
	Tcl_Close(interp, chan);
	Tcl_DStringFree(&dstring);
	return TCL_OK;
    }
    tkimg_PhotoExpand(imageHandle, interp, destX + width, destY + height);

    maxintensity = strtoul(p, &p, 0);
    if ((type != '4') && !maxintensity) {
	Tcl_DStringSetLength(&dstring, 0);
	Tcl_Gets(chan, &dstring);
	p = Tcl_DStringValue(&dstring);
	maxintensity = strtoul(p, &p, 0);
    }
    Tcl_DStringFree(&dstring);
    line3 = (unsigned char *) ckalloc(3 * fileWidth);
    block.pixelSize = 1;
    block.pitch = block.width = width;
    block.height = 1;
    block.offset[0] = 0;
    block.offset[1] = 0;
    block.offset[2] = 0;
    block.offset[3] = 0;
    switch(type) {
	case '4':
	    i = (fileWidth+7)/8;
	    line = (unsigned char *) ckalloc(i);
	    while (srcY-- > 0) {
		Tcl_Read(chan,(char *) line, i);
	    }
	    block.pixelPtr = line3;
	    while (height--) {
	        Tcl_Read(chan, (char *) line, i);
	        for (j = 0; j < width; j++) {
		    line3[j] = ((line[(j+srcX)/8]>>(7-(j+srcX)%8) & 1)) ? 0 : 255;
	        }
		tkimg_PhotoPutBlockTk (interp, imageHandle, &block, destX, destY++, width, 1);
	    }
	    break;
	case '5':
	    line = (unsigned char *) ckalloc(fileWidth);
	    while (srcY-- > 0) {
		Tcl_Read(chan, (char *) line, fileWidth);
	    }
	    block.pixelPtr = line + srcX;
	    while (height--) {
		unsigned char *c = block.pixelPtr;
		Tcl_Read(chan, (char *) line, fileWidth);
		if (maxintensity != 255) {
		    for (j = width; j > 0; j--) {
			*c = (((int)*c) * maxintensity) / 255;
			c++;
		    }
		}
		tkimg_PhotoPutBlockTk(interp, imageHandle, &block, destX, destY++, width, 1);
	    }
	    break;
	case '6':
	    i = 3 * fileWidth;
	    line = NULL;
	    while (srcY-- > 0) {
		Tcl_Read(chan, (char *) line3, i);
	    }
	    block.pixelPtr = line3 + (3 * srcX);
	    block.pixelSize = 3;
	    block.offset[1] = 1;
	    block.offset[2] = 2;
	    while (height--) {
		unsigned char *c = block.pixelPtr;
		Tcl_Read(chan, (char *) line3, i);
		if (maxintensity != 255) {
		    for (j = (3 * width - 1); j >= 0; j--) {
			*c = (((int)*c) * maxintensity) / 255;
			c++;
		    }
		}
		tkimg_PhotoPutBlockTk(interp, imageHandle, &block, destX, destY++, width, 1);
	    }
	    break;
    }
    if (line) {
	ckfree((char *) line);
    }
    ckfree((char *) line3);
    Tcl_Close(interp, chan);
    Tcl_ResetResult(interp);
    return TCL_OK;
#else
    Tcl_AppendResult(interp, "Cannot read postscript file: not implemented",
	    (char *) NULL);
    return TCL_ERROR;
#endif
}

static int
ChnWrite(interp, filename, format, blockPtr)
    Tcl_Interp *interp;
    CONST char *filename;
    Tcl_Obj *format;
    Tk_PhotoImageBlock *blockPtr;
{
    Tcl_Channel chan;
    tkimg_MFile handle;
    int result;

    chan = tkimg_OpenFileChannel(interp, filename, 0644);
    if (!chan) {
	return TCL_ERROR;
    }

    handle.data = (char *) chan;
    handle.state = IMG_CHAN;

    result = CommonWrite(interp, &handle, format, blockPtr);
    if (Tcl_Close(interp, chan) == TCL_ERROR) {
	return TCL_ERROR;
    }
    return result;
}

static int
StringWrite(interp, dataPtr, format, blockPtr)
    Tcl_Interp *interp;
    Tcl_DString *dataPtr;
    Tcl_Obj *format;
    Tk_PhotoImageBlock *blockPtr;
{
    tkimg_MFile handle;
    int result;
    Tcl_DString data;
    tkimg_FixStringWriteProc(&data, &interp, &dataPtr, &format, &blockPtr);
    tkimg_WriteInit(dataPtr, &handle);
    result = CommonWrite(interp, &handle, format, blockPtr);
    tkimg_Putc(IMG_DONE, &handle);
    if ((result == TCL_OK) && (dataPtr == &data)) {
	Tcl_DStringResult(interp, dataPtr);
    }
    return result;
}

static int
CommonWrite(interp, handle, format, blockPtr)
    Tcl_Interp *interp;
    tkimg_MFile *handle;
    Tcl_Obj *format;
    Tk_PhotoImageBlock *blockPtr;
{
    return TCL_OK;
}


static int
ChnMatchBeta (interp, chan, fileName, format, widthPtr, heightPtr) /* PDF */
    Tcl_Interp *interp;
    Tcl_Channel chan;
    CONST char *fileName;
    Tcl_Obj *format;
    int *widthPtr, *heightPtr;
{
    tkimg_MFile handle;

    tkimg_FixChanMatchProc(&interp, &chan, &fileName, &format, &widthPtr, &heightPtr);

    handle.data = (char *) chan;
    handle.state = IMG_CHAN;

    return CommonMatchPDF(&handle, format, widthPtr, heightPtr);
}

static int
ObjMatchBeta(interp, data, format, widthPtr, heightPtr) /* PDF */
    Tcl_Interp *interp;
    Tcl_Obj *data;
    Tcl_Obj *format;
    int *widthPtr, *heightPtr;
{
    tkimg_MFile handle;

    tkimg_FixObjMatchProc(&interp, &data, &format, &widthPtr, &heightPtr);

    if (!tkimg_ReadInit(data, '%', &handle)) {
	return 0;
    }

    return CommonMatchPDF(&handle, format, widthPtr, heightPtr);
}

static int
CommonMatchPDF(handle, format, widthPtr, heightPtr)
    tkimg_MFile *handle;
    Tcl_Obj *format;
    int *widthPtr, *heightPtr;
{
    unsigned char buf[41];
    int zoomx, zoomy, w, h;

    if ((tkimg_Read(handle, (char *) buf, 5) != 5)
	    || (strncmp("%PDF-", (char *) buf, 5) != 0)) {
	return 0;
    }

    /* Here w and h should be set to the bounding box of the pdf
     * data. But I don't know how to extract that from the file.
     * For now I just assume A4-size with 72 pixels/inch. If anyone
     * has a better idea, please mail to <j.nijtmans@chello.nl>.
     */

    w = 612/10;
    h = 792/10;

    if (parseFormat(format, &zoomx, &zoomy) >= 0) {
	w = (w * zoomx + 36) / 72;
	h = (h * zoomy + 36) / 72;
    }
    if ((w <= 0) || (h <= 0)) return 0;
    *widthPtr = w;
    *heightPtr = h;
    return 1;
}
